activerecord/lib/active_record/insert_all.rb

```diff
@@ -137,28 +135,31 @@ def values_list
         end
 
         def returning
-          quote_columns(insert_all.returning).join(",") if insert_all.returning
+          format_columns(insert_all.returning) if insert_all.returning
         end
 
         def conflict_target
           if index = insert_all.unique_by
-            sql = +"(#{quote_columns(index.columns).join(',')})"
+            sql = +"(#{format_columns(index.columns)})"
             sql << " WHERE #{index.where}" if index.where
             sql
           elsif update_duplicates?
-            "(#{quote_columns(insert_all.primary_keys).join(',')})"
+            "(#{format_columns(insert_all.primary_keys)})"
           end
         end
 
-        def updatable_columns
-          quote_columns(insert_all.updatable_columns)
+        def upsert_sql(&block)
+          case insert_all.upsert
+          when true
+            quote_columns(insert_all.updatable_columns).map(&block).join(",")
```

Comment:
@palkan https://github.com/rails/rails/pull/35636 would essentially be cut down to just extending this with:

```ruby
when Array
  quote_columns(insert_all.upsert).map(&block).join(",")
when String
  insert_all.upsert
end
```